core/dbt/clients/registry.py

```diff
@@ -80,7 +98,12 @@ def package(name, registry_base_url=None):
 
 
 def package_version(name, version, registry_base_url=None):
-    return _get_with_retries("api/v1/{}/{}.json".format(name, version), registry_base_url)
+    # returns a dictionary of metadata for a single version of a package
+    expected_type = dict
+    expected_keys = ["name", "packages", "downloads"]
+    return _get_with_retries(
+        "api/v1/{}/{}.json".format(name, version), registry_base_url, expected_type, expected_keys
+    )
```

Comment:
We can avoid a lot of superfluous requests by making this a filter on the `package` request response, rather than a separate request. This is what I'm trying over in https://github.com/dbt-labs/dbt-core/pull/4983:

https://github.com/dbt-labs/dbt-core/blob/25d4c96d86bd04d9bf271e8e18d5a0b8ed8295e5/core/dbt/clients/registry.py#L85-L87

But we're not validating `["name", "packages", "downloads"]` as expected keys of the `package` response there, just the presence of a `versions` key (with a list of versions).

Maybe we should take a hybrid approach like:
- Grab the specific version info from the `package` response
- Check to ensure that the expected keys are present (`["name", "packages", "downloads"]`)
- If they're missing for some reason, fall back to making the current version-specific request, with retries: "api/v1/{}/{}.json"`

Reply:
> Check to ensure that the expected keys are present (["name", "packages", "downloads"])

Another way of doing this: `RegistryPackageMetadata.from_dict()` here, on the specific version response, instead of in `dbt.deps.registry._fetch_metadata`